---
title: "PhD_toolbox_final_exercise"
author: "Stefania"
date: "April 6, 2017"
output: html_document

---

Rationale of the experiment:

The genome of 18 Oidiodendron maius strains has been sequenced. 9 isolates are metal tolerant (T) and 9 isolates are metal sensitive (S). The mapping has been done using the already available and annotated genome of the strain "Zn" as reference. All the uncovered positions in the genomes of the other isolates, are regions of the genome of O. maius Zn that are absent in the other genomes. Those regions could have evolved under the pressure of the heavy metal polluted soil in which O. maius Zn has been isolated, and could thus cointain genes functionally related with its metal tolerance. In this exercice I focus on the scaffold 35 of the genome. 

Aim of the exercise : 

Start using the tools we learnt at the course!

Plotting the coverage data of scaffold 35 and observe (a) whether there are differences between T and S in the number of reads covering the scaffold and (b) how the reads map along the scaffold in the different isolates.

###Step 1 : Gather the data from 18 files (one per isolate) to one single file.
The number of reads for each position in scaffold 35 are exported to one single  column. The output file will contain 4 columns: the type of isolate (T or S, col n° 1), the isolate name (col n° 2), the position in the scaffold (col n° 3), and the number of reads per each position (col n° 4).


```{python echo = FALSE}

import os, glob, sys

##The HOME of the data of the project
HOME = "/media/stefania/DATA/sync/Seminari-scuole/2017_Unix_R_basic/Stefania_project_def/Stefania_project_def/data_scaffold35/"

##List of cov scaffold 35 files
cov_sc35 = glob.glob(os.path.join(HOME+"splitted", "*.coverage.bp.sc35.txt"))
#print cov_sc35
print "Source of data:", len(cov_sc35), "files called 'isolate_id'.coverage.bp.sc35.txt"
outfile = os.path.join(HOME+"merged", "all.cov.bp.sc35.txt")
print "output:", outfile
print "Type of isolate  Isolate_id:"
out = open(outfile, "w")
for file in cov_sc35 :
    id = file.split("/")[-1].split(".")[0]
    #print id
    if id == "4E" or id == "A" or id == "E" or id == "13G" or id == "1354" or id == "1357" or id == "1358" or id == "91" or id == "5L3" :
            print "S", id
    else :                
            print "T", id
    input = open(file, "r")
    for line in input.readlines():
         cols = line.split("\t")
        
        
         if id == "4E" or id == "A" or id == "E" or id == "13G" or id == "1354" or id == "1357" or id == "1358" or id == "91" or id == "5L3" :
                out.write("S"+"\t"+id+"\t"+cols[3]+"\t"+cols[4])
               
         else :      
            out.write("T"+"\t"+id+"\t"+cols[3]+"\t"+cols[4])
           
            
out.close()

```

###Step 2: Load the libraries and import the data 

Note that the `echo = FALSE` parameter was included to the code chunks in order not to print the R code that generated the plots.

```{r echo = FALSE}
#install tidyverse package (once) 
#install.packages("tidyverse")

#load the libraries
library(tidyverse)
library(gridExtra)

#import data and read
cov_sc35  <- read.csv("/media/stefania/DATA/sync/Seminari-scuole/2017_Unix_R_basic/Stefania_project_def/Stefania_project_def/data_scaffold35/merged/all.cov.bp.sc35.txt", header=FALSE, sep="\t", stringsAsFactors = FALSE)

#change columns names
colnames(cov_sc35)[1] <- "type"
colnames(cov_sc35)[2] <- "isolate"
colnames(cov_sc35)[3] <- "pos_bp"
colnames(cov_sc35)[4] <- "n_of_reads"

#check imported data
#is.vector("n_of_reads")
#is.vector("type")
#dim(cov_sc35)
#colnames(cov_sc35)
```

###Step 3: (a) How many reads cover scaffold 35 in S vs T isolates?

```{r echo = FALSE}
##group by "type" and create the background
reads_counts <- cov_sc35 %>%
  group_by(type, isolate) %>%
  summarize(sum_reads = sum(n_of_reads))

#add axes layer
sc35_cov_plot <- ggplot(data = reads_counts,
                        aes(x = type, y = sum_reads))
#add graph layer: total n° of reads in S isolates vs T isolates
sc35_cov_plot + geom_boxplot(aes(group = type)) + 
  geom_point(aes(group = type, color = isolate, shape = isolate), size = 4) +
  scale_shape_manual(values = c(0:17)) +
  labs(title = 'Total reads mapping on scaffold 35',
       x = 'Type of isolate', y = 'Total n° of reads') +
  theme_minimal()

```

Fig. 1 : Number of reads mapping scaffold 35 in each isolate and distribution of the values for sensitive (S) and tolerant (T) isolates.

###Step 3: (b) How are the reads distributed along the scaffold 35 in the different isolates? Are there completely uncovered bases?


```{r echo = FALSE}
#Group by isolate and by position in the scaffold
#filter by type of isolate, S first, T later, in order to obtain separate graphics.
reads_pos_S <- cov_sc35 %>%
  filter(type == "S") %>%
  group_by(pos_bp, isolate, type) %>%
  summarize(pos_reads = n_of_reads)
 
reads_pos_T <- cov_sc35 %>%
  filter(type == "T") %>%
  group_by(pos_bp, isolate, type) %>%
  summarize(pos_reads = n_of_reads)

#plot the number of reads per each position in each isolate

##S isolates
sc35_cov_pos_plot_S <- ggplot(data = reads_pos_S,
                              aes(x = pos_bp, y = pos_reads))+
  geom_line(aes(group = isolate, color = isolate)) + 
  facet_grid(isolate ~., scales = "free_y") +
  labs(title = 'Sensitive isolates',
       x = 'Position in scaffold 35 (bp)',
       y = '') +
  theme(axis.text.x = element_text(colour="grey20", size=8,
                                   angle=0, hjust=.5, vjust=.5),
        axis.text.y = element_text(colour="grey20", size=8),
        text=element_text(size=12, family="Arial")) +
  guides(color = "none")
  

##T isolates
sc35_cov_pos_plot_T <- ggplot(data = reads_pos_T,
                              aes(x = pos_bp, y = pos_reads)) + 
  geom_line(aes(group = isolate, color = isolate)) + 
  facet_grid(isolate ~., scales = "free_y", as.table = FALSE) +
  labs(title = 'Tolerant isolates',
       x = 'Position in scaffold 35 (bp)',
       y = 'n° of reads') +
  theme(axis.text.x = element_text(colour="grey20", size=8,
                                   angle=0, hjust=.5, vjust=.5),
        axis.text.y = element_text(colour="grey20", size=8),
        text=element_text(size=12, family="Arial")) +
  guides(color = "none")
```


```{r echo = FALSE}
##combine the two plots (T and S) in one single figure.
grid.arrange(sc35_cov_pos_plot_T,  sc35_cov_pos_plot_S, ncol=2)
```

Fig. 2: Mapping of the reads along scaffold 35 in each isolate. 

###Conclusions
The coverage of scaffold 35 shows a higher variability in the tolerant than the sensitive isolates. There are regions of the genome of O. maius Zn that are completely missing in many other isolates, both tolerant and sensitive. Next step will be finding the genes annotated in those regions. 



